keps/sig-auth/2579-psp-replacement/README.md

```diff
@@ -0,0 +1,1086 @@
+
+```
+apiVersion: apiserver.config.k8s.io/v1
+kind: AdmissionConfiguration
+plugins:
+- name: PodSecurity
+  configuration:
+    defaults:  # Defaults applied when a mode label is not set.
+      enforce:         <default enforce policy level>
+      enforce-version: <default enforce policy version>
+      audit:         <default audit policy level>
+      audit-version: <default audit policy version>
+      warn:          <default warn policy level>
+      warn-version:  <default warn policy version>
+    exemptions:
+      usernames:         [ <array of authenticated usernames to exempt> ]
+      runtimeClassNames: [ <array of runtime class names to exempt> ]
+      namespaces:        [ <array of namespaces to exempt> ]
+...
+```
+
+[Admission Configuration file]: https://github.com/kubernetes/kubernetes/blob/3d6026499b674020b4f8eec11f0b8a860a330d8a/staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1/types.go#L27
+
+#### Defaulting
+
+The default policy level and version for each mode (when no label is present) can be statically
+configured. The default for the static configuration is `privileged` and `latest`.
+
+#### Exemptions
+
+Policy exemptions can be statically configured. Exemptions must be explicitly enumerated, and don’t
+support indirection such as label or group selectors. Requests meeting criteria are ignored by the
+admission controller (enforce, audit and warn) except for recording an [audit
+annotation](#audit-annotations). Exemption dimensions include:
+
+- Usernames: requests from users with an exempt authenticated (or impersonated) username are ignored.
+- RuntimeClassNames: pods and [templated pods] with specifying an exempt runtime class name are ignored.
+- Namespaces: pods and [templated pods] in an exempt namespace are ignored.
+
+The username exemption is special in that the creating user is not persisted on the pod object, and
+the pod may be modified by different non-exempt users in the future. See [Updates](#updates) for
+details on how non-exempt updates of a previously exempted pod are handled. Use cases for username
+exemptions include:
+
+- Trusted controllers that create pods in tenant namespaces with additional 3rd party enforcement on
+  the privileged pods.
+- Break-glass operations roles, for example for [debugging workloads in a restricted
+  namespace](#ephemeral-containers).
+
+### Risks and Mitigations
+
+**Future proofing:** The policy versioning aspects of this proposal are designed to anticipate
+breaking changes to policies, either in light of new threats or new knobs added to pods. However, if
+there is a new feature that needs to be restricted but doesn't have a sensible hardcoded
+requirement, we would be put in a hard place. Hopefully the adoption of this proposal will
+discourage such fields from being added.
+
+**Scope creep:** How do we decide which fields are in-scope for policy restrictions? There are a
```

Comment:
This risk lacks a mitigation. Did you have one in mind? Can we scope to some definition of "privilege escalation" and exclude run of the mill info-leaks and dos vectors from baseline and restricted?

Reply:
I'm not sure I'm ready to rule out local DoS mitigations from this. I would consider restricting this to runtime controls, for example excluding fields that only affect the scheduling (nodeSelector, ~tolerations, topology constraints) or control plane (labels, owner references, DeletionGracePeriodSeconds). I think we could also rule out all of PodStatus, since end users shouldn't be granted write to status. Networking gets a bit fuzzier, since we already include HostPorts, HostNetwork, and capabilities...

WDYT? @liggitt anything to add?

Reply:
Documented the scope restrictions here.

Reply:
Agree on pod status being out of scope. I'd say that anything scheduling/topology related is out of scope, as is anything quota/count/quantity-related ("you may use X, but only Y many of them")